akka-stream/src/main/scala/akka/stream/impl/Sources.scala

```diff
@@ -15,124 +15,167 @@ import scala.compat.java8.FutureConverters._
             val bufferOverflowException = new BufferOverflowException(s"Buffer overflow (max capacity was: $maxBuffer)!")
-            promise.success(QueueOfferResult.Failure(bufferOverflowException))
+            offer.promise.success(QueueOfferResult.Failure(bufferOverflowException))
             completion.failure(bufferOverflowException)
             failStage(bufferOverflowException)
           case Backpressure ⇒
             pendingOffer match {
               case Some(_) ⇒
-                promise.failure(new IllegalStateException("You have to wait for previous offer to be resolved to send another request"))
+                offer.promise.failure(new IllegalStateException("You have to wait for previous offer to be resolved to send another request"))
               case None ⇒
-                pendingOffer = Some((elem, promise))
+                pendingOffer = Some(offer)
             }
         }
       }
 
-      private val callback: AsyncCallback[(T, Offered)] = getAsyncCallback(tuple ⇒ {
-        val (elem, promise) = tuple
-
-        if (maxBuffer != 0) {
-          bufferElem(elem, promise)
-          if (pulled) {
-            push(out, buffer.dequeue())
-            pulled = false
+      private val callback: AsyncCallback[Input[T]] = getAsyncCallback {
+        case offer @ Offer(elem, promise) ⇒
+          if (maxBuffer != 0) {
+            bufferElem(offer)
+            if (isAvailable(out)) push(out, buffer.dequeue())
+          } else if (isAvailable(out)) {
+            push(out, elem)
+            promise.success(QueueOfferResult.Enqueued)
+          } else if (pendingOffer.isEmpty) pendingOffer = Some(offer)
+          else overflowStrategy match {
+            case DropHead | DropBuffer ⇒
+              pendingOffer.get.promise.success(QueueOfferResult.Dropped)
+              pendingOffer = Some(offer)
+            case DropTail | DropNew ⇒
+              promise.success(QueueOfferResult.Dropped)
+            case Fail ⇒
+              val bufferOverflowException = new BufferOverflowException(s"Buffer overflow (max capacity was: $maxBuffer)!")
+              promise.success(QueueOfferResult.Failure(bufferOverflowException))
+              completion.failure(bufferOverflowException)
+              failStage(bufferOverflowException)
+            case Backpressure ⇒
+              promise.failure(new IllegalStateException("You have to wait for previous offer to be resolved to send another request"))
+          }
+        case Completion ⇒
+          if (maxBuffer != 0 && buffer.nonEmpty || pendingOffer.nonEmpty) terminating = true
+          else {
+            completion.success(Done)
+            completeStage()
           }
-        } else if (pulled) {
-          push(out, elem)
-          pulled = false
-          promise.success(QueueOfferResult.Enqueued)
-        } else pendingOffer = Some(tuple)
-      })
+        case Failure(ex) ⇒
+          completion.failure(ex)
+          failStage(ex)
+      }
```

Comment:
I would have liked a bit less cyclomatic complexity here, can't it be split into smaller parts?